node/cdn/src/blocks.rs

```diff
@@ -138,143 +146,192 @@ pub async fn load_blocks<N: Network>(
-            let client_clone = client.clone();
-            let base_url_clone = base_url.to_string();
-            let failed_clone = failed.clone();
-            handle_dispatch_error(move || {
-                let ctx = ctx.clone();
-                let client = client_clone.clone();
-                let base_url = base_url_clone.clone();
-                let failed = failed_clone.clone();
-                async move {
-                    // If the sync failed, return with an empty vector.
-                    if failed.read().is_some() {
-                        return std::future::ready(Ok(vec![])).await
-                    }
                     // Prepare the URL.
-                    let blocks_url = format!("{base_url}/{start}.{end}.blocks");
-                    // Fetch the blocks.
-                    let blocks: Vec<Block<N>> = match cdn_get(client, &blocks_url, &ctx).await {
-                        Ok(blocks) => blocks,
-                        Err(error) => {
-                            error!("Failed to request {ctx} - {error}");
-                            failed.write().replace(error);
-                            return std::future::ready(Ok(vec![])).await
+                    let blocks_url = format!("{base_url_clone}/{start}.{end}.blocks");
+                    let ctx = format!("blocks {start} to {end}");
+                    // Download blocks, retrying on failure.
+                    let mut attempts = 0;
+                    let request_time = Instant::now();
+
+                    loop {
+                        // Fetch the blocks.
+                        match cdn_get(client_clone.clone(), &blocks_url, &ctx).await {
+                            Ok::<Vec<Block<N>>, _>(blocks) => {
+                                // Keep the collection of pending blocks sorted by the height.
+                                let mut pending_blocks = pending_blocks_clone.lock();
+                                for block in blocks {
+                                    match pending_blocks.binary_search_by_key(&block.height(), |b| b.height()) {
+                                        Ok(_idx) => warn!("Duplicate pending block at height {}", block.height()),
+                                        Err(idx) => pending_blocks.insert(idx, block),
+                                    }
+                                }
+                                debug!(
+                                    "Received {ctx} in {:.2?} ({} queued for insertion)",
+                                    request_time.elapsed(),
+                                    pending_blocks.len()
+                                );
+                                break;
+                            }
+                            Err(error) => {
+                                // Increment the attempt counter, and wait with a linear backoff.
+                                attempts += 1;
```

Comment:
should we have some max attempts?

Reply:
happy to impose some limit; @howardwu would you like to suggest one?

Reply:
I introduced a limit of 10 attempts for now; if it is breached, the sync gets aborted.